File: iOS/PricetarPartner/PricetarPartner/Http/Requester/FetchChatGroupRequester.swift
//
//  FetchChatGroupRequester.swift
//  PricetarPartner
//
//  Created by Leapfrog-Inc. on 2021/03/02.
//

import Foundation

class FetchChatGroupRequester {
    
    static let shared = FetchChatGroupRequester()
    
    var dataList = [ChatGroupData]()
    
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        var params = ["command": "getChatGroup"]
        params["userId"] = SaveData.shared.userId
        
        ApiRequester.post(params: params) { [weak self] result, data in
            if result, let dic = data as? Dictionary<String, Any> {
                if (dic["result"] as? String) == "0", let users = (dic["chatGroups"] as? Array<Dictionary<String, Any>>) {
                    self?.dataList = users.compactMap { ChatGroupData(data: $0) }
                    completion(true)
                    return
                }
            }
            completion(false)
        }
    }
}

File: Android/app/src/main/java/jp/co/lfg/pricetarpartner/Http/Requester/FetchChatGroupRequester.java
package jp.co.lfg.pricetarpartner.Http.Requester;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;

import jp.co.lfg.pricetarpartner.Http.ApiRequester;
import jp.co.lfg.pricetarpartner.Http.DataModel.ChatGroupData;
import jp.co.lfg.pricetarpartner.Http.DataModel.UserData;
import jp.co.lfg.pricetarpartner.System.SaveData;

public class FetchChatGroupRequester {

    private static FetchChatGroupRequester mRequester = new FetchChatGroupRequester();

    private FetchChatGroupRequester() {}

    public static FetchChatGroupRequester getInstance() {
        return mRequester;
    }

    public ArrayList<ChatGroupData> dataList = new ArrayList<>();

    public void fetch(final Callback callback) {

        HashMap<String, String> params = new HashMap<>();
        params.put("command", "getChatGroup");
        params.put("userId", SaveData.getInstance().userId);

        ApiRequester.post(params, new ApiRequester.Callback() {
            @Override
            public void didReceiveData(boolean result, JSONObject json) {
                if (result) {
                    try {
                        String apiResult = json.getString("result");
                        if (apiResult.equals("0")) {
                            ArrayList<ChatGroupData> chatGroups = new ArrayList<>();
                            JSONArray chatGroupArray = json.getJSONArray("chatGroups");
                            for (int i = 0; i < chatGroupArray.length(); i++) {
                                ChatGroupData chatGroupData = ChatGroupData.create(chatGroupArray.getJSONObject(i));
                                if (chatGroupData != null) {
                                    chatGroups.add(chatGroupData);
                                }
                            }
                            dataList = chatGroups;
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch (JSONException e) {}
                }
                callback.didReceiveData(false);
            }
        });
    }

    public interface Callback {
        void didReceiveData(boolean result);
    }
}
